mint-client/src/clients/user.rs

```diff
@@ -379,6 +367,113 @@ impl UserClient {
             .await
             .map_err(|_| ClientError::WaitContractTimeout)?
     }
+
+    /// Fetches the TransactionStatus for a txid
+    /// Polling should *only* be set to true if it is anticipated that the txid is valid but has not yet been processed
+    pub async fn fetch_tx_outcome(
+        &self,
+        tx: TransactionId,
+        polling: bool,
+    ) -> Result<TransactionStatus, ClientError> {
+        // did not choose to use the MintClientError is_retryable logic because the 404 error should normaly
+        // not be retryable just in this specific case...
+        let status;
+        loop {
+            match self.context.api.fetch_tx_outcome(tx).await {
+                Ok(s) => {
+                    status = s;
+                    break;
+                }
+                Err(_e) if polling => tokio::time::sleep(Duration::from_secs(1)).await,
+                Err(e) => return Err(ClientError::MintClientError(MintClientError::ApiError(e))),
+            }
+        }
+        Ok(status)
+    }
+
+    async fn sign_and_submit_tx<R: RngCore + CryptoRng>(
+        &self,
+        coin_keys: &[KeyPair],
+        inputs: Vec<mint_tx::Input>,
+        outputs: Vec<mint_tx::Output>,
+        mut rng: R,
+    ) -> Result<TransactionId, ClientError> {
+        let txid = mint_tx::Transaction::tx_hash_from_parts(&inputs, &outputs);
+
+        let signature = minimint::transaction::agg_sign(
+            coin_keys,
+            txid.as_hash(),
+            &self.context.secp,
+            &mut rng,
+        );
+
+        let transaction = mint_tx::Transaction {
+            inputs,
+            outputs,
+            signature: Some(signature),
+        };
+        let tx_id = transaction.tx_hash();
+
+        let mint_tx_id = self.context.api.submit_transaction(transaction).await?;
+
+        assert_eq!(
+            tx_id, mint_tx_id,
+            "Federation is faulty, returned wrong tx id."
+        );
+
+        Ok(tx_id)
+    }
+
+    pub async fn make_change<R: RngCore + CryptoRng>(
+        &self,
+        coins: Coins<SpendableCoin>,
+        amount: Amount,
+        mut rng: R,
+    ) -> Result<(), ClientError> {
```

Comment:
I'm not sure what to call this function or whether to just leave this logic inside `select_and_spend_coins`?